modules/core/lib/Auth/Process/PairwiseID.php

```diff
@@ -0,0 +1,134 @@
+<?php
+
+declare(strict_types=1);
+
+namespace SimpleSAML\Module\core\Auth\Process;
+
+use Exception;
+use SAML2\Constants;
+use SAML2\XML\saml\NameID;
+use SimpleSAML\Assert\Assert;
+use SimpleSAML\Auth;
+use SimpleSAML\Utils;
+
+/**
+ * Filter to generate the Pairwise ID attribute.
+ *
+ * See: http://docs.oasis-open.org/security/saml-subject-id-attr/v1.0/csprd01/saml-subject-id-attr-v1.0-csprd01.html
+ *
+ * By default, this filter will generate the ID based on the UserID of the current user.
+ * This is generated from the attribute configured in 'identifyingAttribute' in the
+ * authproc-configuration.
+ *
+ * Example - generate from attribute:
+ * <code>
+ * 'authproc' => [
+ *   50 => [
+ *       'core:PairwiseID',
+ *       'identifyingAttribute' => 'uid',
+ *       'scope' => 'example.org',
+ *   ]
+ * ]
+ * </code>
+ *
+ * @package SimpleSAMLphp
+ */
+class PairwiseID extends Auth\ProcessingFilter
+{
+    /**
+     * The attribute we should generate the pairwise id from.
+     *
+     * @var string
+     */
+    private string $identifyingAttribute;
+
+    /**
+     * The scope to use for this attribute.
+     *
+     * @var string
+     */
+    private string $scope;
+
+    /**
+     * @var \SimpleSAML\Utils\Config|string
+     * @psalm-var \SimpleSAML\Utils\Config|class-string
+     */
+    protected $configUtils = Utils\Config::class;
+
+
+    /**
+     * Initialize this filter.
+     *
+     * @param array &$config  Configuration information about this filter.
+     * @param mixed $reserved  For future use.
+     */
+    public function __construct(array &$config, $reserved)
+    {
+        parent::__construct($config, $reserved);
+
+        Assert::keyExists($config, 'identifyingAttribute', "Missing mandatory 'identifyingAttribute' config setting.");
+        Assert::keyExists($config, 'scope', "Missing mandatory 'scope' config setting.");
+        Assert::stringNotEmpty($config['identifyingAttribute']);
+        Assert::regex(
+            $config['scope'],
+            '/^[a-zA-Z0-9.-]+$/i',
+            'PairwiseID: \'scope\' contains illegal characters.'
```

Comment:
If we want to be even more sure we do not release anything spec incompliant we change the regex to `/^[a-z0-9][a-z0-9.-]{0,126}$/i` (change the exception message to match). This might be good to do since one problem with the existing identifiers is the amount of variance between implementations. So be strict in what we can send.

Reply:
I'd say we bump the specs a little but to {1,126}, because without a scope this can _never_ be a globally unique ID.. It also ensures a non-empty string

Reply:
Ah, I guess we were both wrong on the specs..  I think I got them right now!

Reply:
I'm not sure where my regex was wrong? I had at least a length of one? But agreed that probably any length < 4 is an error and cannot be globally unique.

Reply:
I misread it.. it wasn't all wrong, except for the fact that you're missing the A-Z capitals..
My suggestion would be to use the {0,126} to enforce the pattern and log a warning if it's not matching at least {3,126}..

Reply:
I'm not missing the capitals because the regexp uses the `i` modifier.